Add upstream Ansible collection generation and Galaxy publishing from profile playbooks

ggbecker opened the pull request:
#### Description:

- Add `utils/ansible_roles_to_collection.py` to bundle profile Ansible roles into a
  publishable `redhatofficial.rhel_hardening_roles` Ansible collection. Vendors
  `community.general.ini_file` and `ansible.posix` modules (`firewalld`, `mount`,
  `seboolean`, `sysctl`) into `plugins/modules/` and rewrites all FQCN references.
  Uses `ansible-galaxy collection build` to produce a Galaxy-compatible artifact.
  Accepts multiple `--roles-dir` flags to merge roles from separate per-version builds
  (the expected downstream flow). Collection metadata (`--description`, `--documentation`,
  `--homepage`, `--issues`) is overridable. Version is derived from `CMakeLists.txt`.
- Extend `ansible_playbook_to_role.py` with `--local-roles-dir` to cache roles locally
  alongside the Galaxy push. Fix `--product` being silently ignored in dry-run mode.
- Add CMake build integration: `SSG_ANSIBLE_ROLES_ENABLED` and
  `SSG_ANSIBLE_COLLECTION_ENABLED` as `cmake_dependent_option` entries. Fix
  `ssg_build_ansible_roles` stamp file creation and `DEPENDS` chain; depend on
  `${PRODUCT}` instead of `${PRODUCT}-content`. The collection is the only installed
  artifact (no standalone roles install).
- Document the collection build and manual Galaxy upload flow in `docs/release_process.md`.

#### Rationale:

- Vendoring `ansible.posix` and `community.general` removes them as runtime dependencies
  for collection consumers.

#### Review Hints:

- Build and verify the collection locally:
  ```bash
  ADDITIONAL_CMAKE_OPTIONS="-DSSG_ANSIBLE_ROLES_ENABLED=TRUE" \
      ./build_product rhel9 --datastream
  ninja -C build generate-rhel9-ansible-roles
  python3 utils/ansible_roles_to_collection.py \
      --roles-dir build/ansible_roles \
      --output-dir /tmp/test-collection \
      --build
  grep -r "community\.general\.\|ansible\.posix\." \
      /tmp/test-collection/ansible_collections/redhatofficial/rhel_hardening_roles/roles/ \
      && echo FAIL || echo OK
  ```
- Key files: `utils/ansible_roles_to_collection.py`, `cmake/SSGCommon.cmake`,
  `docs/release_process.md`

ggbecker committed:
Move pygithub import right before when it is needed.

We need a way to run in dry run mode without requiring the pygithub module.

This way also helps the ruff upstream gating code.

ggbecker committed:
Add a new SSG_ANSIBLE_ROLES_ENABLED to build ansible roles.

Introduces a new macro function to build ansible roles out of the
profile ansible playbooks. These files can be further turned into
Ansible collections for example to support ansible content publishing.

ggbecker committed:
Do not print information that pollutes the build system output.

ggbecker commented on utils/ansible_playbook_to_role.py:
```diff
@@ -574,6 +566,12 @@ def main():
             PlaybookToRoleConverter(playbook_full_path).save_to_disk(args.dry_run)
     else:
+        try:
+            from github import Github, InputGitAuthor, UnknownObjectException
```
I've reworked how the imports are being handled. Hopefully it will make this test to pass again.

ggbecker committed:
Switch Ansible Roles org name to CaC.

ggbecker committed:
Dry run mode overrides the profile and product allow list.

utils/ansible_playbook_to_role.py: When running in dry mode, the script
will process every available ansible playbook found in the build/ansible
folder.

ggbecker commented:
This is now parked, but it is still valid. Let's keep it open for now.

ggbecker committed:
Add upstream Ansible collection generation from roles.

Introduces utils/ansible_roles_to_collection.py, a standalone script
that bundles ComplianceAsCode Ansible roles into a publishable Ansible
collection for Ansible Galaxy.

The script vendors modules from community.general (ini_file) and
ansible.posix (firewalld, mount, seboolean, sysctl) directly into the
collection's plugins/modules/ directory, eliminating runtime dependencies
on those external collections. FQCN references in role task files are
rewritten to point at the vendored copies inside the new collection.
Only roles for products in PRODUCT_ALLOWLIST (rhel8, rhel9, rhel10) are
included, matching the same filter used by ansible_playbook_to_role.py.

The collection version is read automatically from CMakeLists.txt so it
always matches the SSG release version without manual input.

Publishing to Ansible Galaxy is done by passing --galaxy-token, which
uploads the built .tar.gz artifact via the Galaxy API. There is no GitHub
repository sync for collections (unlike roles); the tarball upload is the
only supported publishing mechanism for collections.

ansible_playbook_to_role.py gains --local-roles-dir so that roles are
saved locally alongside the GitHub push during the release process. The
saved directory is then passed directly to ansible_roles_to_collection.py
without needing to run the script a second time.

CMake integration is added via the SSG_ANSIBLE_COLLECTION_ENABLED option,
which creates an ansible-collection build target that runs after all
product roles have been generated. The collection is versioned using the
SSG_VERSION and installed under ${SSG_ANSIBLE_COLLECTION_INSTALL_DIR}.

docs/release_process.md documents the full manual release flow for the
collection as a follow-up step to the existing Ansible roles section.

ggbecker commented:
I have revamped this pull request which now builds the ansible collections out of the ansible roles that can be published on ansible galaxy

ggbecker commented:
We agreed on bundling the used community/posix ansible modules even in the upstream code.

Mab879 commented on cmake/SSGCommon.cmake:
```diff
@@ -257,6 +257,56 @@ macro(ssg_build_ansible_playbooks PRODUCT)
+macro(ssg_build_ansible_roles)
+    set(ANSIBLE_ROLES_DIR "${CMAKE_BINARY_DIR}/ansible_roles")
+    add_custom_command(
+        OUTPUT "${ANSIBLE_ROLES_DIR}/ansible_roles-${PRODUCT}"
```
Please ensure that this stamp file is actually created.

Mab879 commented on cmake/SSGCommon.cmake:
```diff
@@ -257,6 +257,56 @@ macro(ssg_build_ansible_playbooks PRODUCT)
+                --version "${VERSION}"
+                --build
+            COMMAND "${CMAKE_COMMAND}" -E touch "${ANSIBLE_COLLECTION_STAMP}"
+            DEPENDS "${ANSIBLE_ROLES_DIR}"
```
Depending on directory isn't best practice for CMake.

Mab879 commented on utils/ansible_playbook_to_role.py:
```diff
@@ -511,14 +522,14 @@ def locally_clone_and_init_repositories(organization, repo_list):
             # the format is product-playbook-profile.yml
             product, _, profile = root.split("-", 2)
-            if product in product_allowlist and profile not in profile_denylist:
+            if dry_run or (product in product_allowlist and profile not in profile_denylist):
```
--product flag is completely ignored in dry-run mode

Is this expected?

Mab879 commented on utils/test_ansible_collection.sh:
```diff
@@ -0,0 +1,65 @@
+
+echo ""
+echo "=== Step 4: Verify output ==="
+COLLECTION_DIR="${OUTPUT_DIR}/ansible_collections/redhat/rhel_hardening_roles"
```
```suggestion
COLLECTION_DIR="${OUTPUT_DIR}/ansible_collections/redhatoffical/rhel_hardening_roles"
```

Mab879 commented on cmake/SSGCommon.cmake:
```diff
@@ -257,6 +257,56 @@ macro(ssg_build_ansible_playbooks PRODUCT)
     endif()
 endmacro()
 
+macro(ssg_build_ansible_roles)
```
```suggestion
macro(ssg_build_ansible_roles PRODUCT)
```

Mab879 commented on cmake/SSGCommon.cmake:
```diff
@@ -833,8 +883,17 @@ macro(ssg_build_product PRODUCT)
+        if(SSG_ANSIBLE_ROLES_ENABLED)
+            ssg_build_ansible_roles(${PRODUCT})
+            add_dependencies(
+                ${PRODUCT}-content
```
```
  Roles are added as a dependency of ${PRODUCT}-content:
  add_dependencies(${PRODUCT}-content generate-${PRODUCT}-ansible-roles)
  But profile playbooks (line 884) and bash scripts (line 903) use
  add_dependencies(${PRODUCT} ...) instead. By attaching to ${PRODUCT}-content,
  role generation blocks the entire content pipeline, unlike analogous targets.
  It should depend on ${PRODUCT}, not ${PRODUCT}-content.
```
Please review this comment from Claude, seems valid

Mab879 commented on cmake/SSGCommon.cmake:
```diff
@@ -257,6 +257,56 @@ macro(ssg_build_ansible_playbooks PRODUCT)
+    add_custom_command(
+        OUTPUT "${ANSIBLE_ROLES_DIR}/ansible_roles-${PRODUCT}"
+        COMMAND env "PYTHONPATH=$ENV{PYTHONPATH}" "${Python_EXECUTABLE}" "${CMAKE_SOURCE_DIR}/utils/ansible_playbook_to_role.py" --dry-run "${ANSIBLE_ROLES_DIR}" --product "${PRODUCT}" --build-playbooks-dir "${CMAKE_BINARY_DIR}/ansible"
+        DEPENDS generate-all-profile-playbooks-${PRODUCT}
```
```
  The macro has DEPENDS generate-all-profile-playbooks-${PRODUCT} (target only).
  The file header (lines 36-38) states: "Any time you use wrapper targets, 
  still add DEPEND(s) on actual files involved." Compare with line 882 which
  correctly lists both the target and the stamp file. Should add
  "${CMAKE_BINARY_DIR}/ansible/all-profile-playbooks-${PRODUCT}".
```
Please view this comment from Claude, seems valid.

Mab879 commented on cmake/SSGCommon.cmake:
```diff
@@ -257,6 +257,56 @@ macro(ssg_build_ansible_playbooks PRODUCT)
+            DEPENDS "${ANSIBLE_COLLECTION_STAMP}"
+        )
+
+        if(SSG_ANSIBLE_PLAYBOOKS_ENABLED)
```
This is mixing playbooks and collections is this expected?

Mab879 commented on utils/ansible_roles_to_collection.py:
```diff
@@ -0,0 +1,606 @@
+    artifact_path = os.path.join(output_dir, artifact_name)
+    print(f"Building collection artifact: {artifact_name}")
+    with tarfile.open(artifact_path, "w:gz") as tar:
+        tar.add(collection_dir, arcname=".")
```
```
  tar.add(collection_dir, arcname=".")

  This creates member paths like ./plugins/modules/ini_file.py and ./galaxy.yml.
  Standard ansible-galaxy collection build produces entries without the ./
  prefix. Additionally, Galaxy expects MANIFEST.json and FILES.json in
  collection tarballs (generated by ansible-galaxy collection build). The script
  produces neither, which will likely cause Galaxy to reject the upload and
  ansible-galaxy collection install to fail. The script should shell out to
  ansible-galaxy collection build rather than using raw tarfile.
```

Some we need verify before merging.

Mab879 commented on cmake/SSGCommon.cmake:
```diff
@@ -257,6 +257,56 @@ macro(ssg_build_ansible_playbooks PRODUCT)
+    set(ANSIBLE_COLLECTION_DIR "${CMAKE_BINARY_DIR}/ansible_collection")
+    set(ANSIBLE_COLLECTION_STAMP "${ANSIBLE_COLLECTION_DIR}/collection-${VERSION}")
+
+    if(NOT SSG_ANSIBLE_ROLES_ENABLED)
```
Consider using `cmake_dependent_option` here like we do in other parts of the project.

Mab879 commented on utils/test_ansible_collection.sh:
```diff
@@ -0,0 +1,65 @@
+#!/bin/bash
+# Test script: build Ansible roles then generate an Ansible collection.
+#
+# Usage: utils/test_ansible_collection.sh [OUTPUT_DIR] [PRODUCT]
```
Doesn't seem this script is used anywhere is that expected?

ggbecker committed:
Address PR review feedback for Ansible collection generation

- Use ansible-galaxy collection build instead of raw tarfile to produce
  a Galaxy-compatible artifact with MANIFEST.json and FILES.json
- Fix ssg_build_ansible_roles macro: add PRODUCT parameter, create stamp
  file with cmake -E touch, add stamp file alongside target in DEPENDS,
  collect stamps in SSG_ANSIBLE_ROLE_STAMPS global property
- Fix ssg_build_ansible_collection: depend on collected role stamps instead
  of the roles directory, drop SSG_ANSIBLE_ROLES_ENABLED warning (enforced
  by cmake_dependent_option), fix install guard to use SSG_ANSIBLE_ROLES_ENABLED
- Replace option() with cmake_dependent_option() for SSG_ANSIBLE_ROLES_ENABLED
  and SSG_ANSIBLE_COLLECTION_ENABLED to enforce dependency ordering
- Change ansible roles add_dependencies target from ${PRODUCT}-content to
  ${PRODUCT}, matching the pattern used by playbooks and bash scripts
- Fix select_roles_to_upload: always apply product/profile filter; dry-run
  bypassed the allowlist so --product was silently ignored
- Document test_ansible_collection.sh in docs/release_process.md under a
  new "Test the Collection Locally" section; fix stale redhat namespace
  in the tarball example

ggbecker committed:
Make collection metadata configurable

Add --description, --documentation, --homepage, and --issues arguments to
ansible_roles_to_collection.py so downstream consumers can supply their own
values without touching the source. The documentation field is omitted from
galaxy.yml when not provided (it has no upstream default).

ggbecker committed:
Support multiple --roles-dir inputs for multi-version collection builds

Allow --roles-dir to be specified more than once so that roles built
separately for each RHEL major version can be merged into a single
collection without a manual pre-merge step.

detect_modules_to_bundle and copy_roles now iterate over a list of
source directories. If the same role name appears in more than one
directory the first occurrence wins and a warning is printed.

The single-directory upstream CMake path is unaffected: one --roles-dir
flag produces a one-element list, which the updated functions handle
identically to before.

ggbecker commented:
@Mab879 I have addressed your feedback and on top of that I have decided for example to remove the ansible roles files installation CMake directives, since we won't need to install files to be able to generate the collection IMO.

We might need to change some things when the time comes for publishing the downstream ansible collections, but that's something hard to anticipate. I believe I covered a fair amount of things that will be needed for downstream, like defining the namespace through an argument, and other fields for the collection metadata.

Mab879 reviewed (approved):
This PR is a great start. We will need make changes as we improve this system but this a good start.
